src/arizona_template_scanner.erl

```diff
@@ -0,0 +1,548 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright 2024 Arizona Framework <contact@arizonaframe.work>
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(arizona_template_scanner).
+-moduledoc false.
+
+%% API functions.
+-export([scan/1]). -ignore_xref([scan/1]).
+
+%% Types
+-export_type([token/0]).
+-export_type([location/0]).
+-export_type([line/0]).
+-export_type([column/0]).
+-export_type([result/0]).
+-export_type([error_reason/0]).
+
+-type state() :: #{
+    line => line(),
+    column => column(),
+    first_column => non_neg_integer(),
+    position => non_neg_integer()
+}.
+-opaque token() :: {open_tag, location(), binary()}
+                 | {attr_key, location(), binary()}
```

Comment:
It `key`, `name`, here?

Reply:
Yes, `binary()` is the name of the key.

Reply:
What I meant is we should rename it as `attr_name`, to be closer to its semantics. I tend to see `key` as an identifier in a set, or list, or map, which is not the case (especially since it's not unique).

Reply:
Makes sense. It is named `name` as we can see [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes).
I'll change this.